src/megaapi_impl.cpp

```diff
@@ -11380,42 +11392,130 @@ bool MegaApiImpl::processTree(Node* node, TreeProcessor* processor, bool recursi
+            && (order < MegaApi::ORDER_NONE || order > MegaApi::ORDER_ALPHABETICAL_DESC))
     {
         return new MegaNodeListPrivate();
     }
 
-    SdkMutexGuard g(sdkMutex);
-
     if (cancelToken && cancelToken->isCancelled())
     {
         return new MegaNodeListPrivate();
     }
 
-    Node *node = client->nodebyhandle(n->getHandle());
-    if (!node)
+    SdkMutexGuard g(sdkMutex);
+
+    if (cancelToken && cancelToken->isCancelled())
     {
         return new MegaNodeListPrivate();
     }
-
-    SearchTreeProcessor searchProcessor(searchString);
-    for (node_list::iterator it = node->children.begin(); it != node->children.end()
-         && !(cancelToken && cancelToken->isCancelled()); )
+    
+    MegaNodeList *nodeList = nullptr;
+    if (n)
     {
-        processTree(*it++, &searchProcessor, recursive, cancelToken);
+        // if node is provided, it will be the parent node of the tree to explore
+        Node *node = client->nodebyhandle(n->getHandle());
+        if (!node)
+        {
+            return new MegaNodeListPrivate();
+        }
+
+        // searchString and nodeType (if provided), are considered in search
+        SearchTreeProcessor searchProcessor(client, searchString, static_cast<MegaApi::nodefiletype_t>(type));
+        for (node_list::iterator it = node->children.begin(); it != node->children.end()
+             && !(cancelToken && cancelToken->isCancelled()); )
+        {
+            processTree(*it++, &searchProcessor, recursive, cancelToken);
+        }
+
+        vector<Node *>& vNodes = searchProcessor.getResults();
+        sortByComparatorFunction(vNodes, order, *client);
+        nodeList = new MegaNodeListPrivate(vNodes.data(), int(vNodes.size()));
     }
+    else
+    {
+        node_vector result;
+        Node *node;
+
+        // Target parameter is only considered if node is not provided
+        if (target < MegaApi::TARGET_INSHARE || target > MegaApi::TARGET_ALL)
+        {
+            return new MegaNodeListPrivate();
+        }
+
+        if (target == MegaApi::TARGET_ROOTNODES || target == MegaApi::TARGET_ALL)
+        {
+            // Search on rootnodes
+            for (unsigned int i = 0; i < (sizeof client->rootnodes / sizeof *client->rootnodes)
+                  && !(cancelToken && cancelToken->isCancelled()); i++)
+            {
+                node = client->nodebyhandle(client->rootnodes[i]);
+                SearchTreeProcessor searchProcessor(client, searchString, static_cast<MegaApi::nodefiletype_t>(type));
+                processTree(node, &searchProcessor, recursive, cancelToken);
+                node_vector& vNodes = searchProcessor.getResults();
+                result.insert(result.end(), vNodes.begin(), vNodes.end());
+            }
+        }
+
+        if (target == MegaApi::TARGET_INSHARE || target == MegaApi::TARGET_ALL)
+        {
+            // Search on inshares
+            MegaShareList *shares = getInSharesList(MegaApi::ORDER_NONE);
```

Comment:
Yes using one of the RAII pointers is much better for cases like this.   About the ::mega prefix, we shouldn't need that when coding inside the mega namespace - there is one exception, we have written our own ::mega::make_unique<> template since that is a very useful C++14 feature, not available in C++11.  thanks

Reply:
The problem here to use `::mega::make_unique<>` is that the object is created inside `getInSharesList()`, which returns a raw pointer, so the performance improvement brought by `make_unique` would require to revamp all usages of `getInSharesList()`

Reply:
Nevertheless, the pointer is now a unique_ptr :)